src/notification-channel.cc

```diff
@@ -0,0 +1,96 @@
+#include "notification-channel.h"
+
+#include "util.h"
+#include "util-inl.h"
+
+namespace node {
+
+NotificationChannel::NotificationChannel(uv_loop_t* loop,
+                                         NotificationCallback callback,
+                                         void* data)
+    : event_loop_(loop),
+      callback_(callback),
+      data_(data) {
+  CHECK_EQ(0, uv_async_init(event_loop(),
+                            notification_signal_async(),
+                            AsyncCallback));
+  ref();
+  notification_signal_async()->data = this;
+  CHECK_EQ(uv_mutex_init(access_mutex()), 0);
+}
+
+void NotificationChannel::Dispose() {
+  ref();
+  {
+    ScopedLock::Mutex lock(access_mutex());
+    is_closing_ = true;
+    callback_pending_ = true;
+  }
+
+  uv_close(reinterpret_cast<uv_handle_t*>(notification_signal_async()),
+           CloseCallback);
+
+  while (!close_callback_called_)
+    uv_run(event_loop(), UV_RUN_ONCE);
+
+  notification_signal_async()->data = nullptr;
+  uv_mutex_destroy(access_mutex());
+}
+
+void NotificationChannel::Ref() {
+  ScopedLock::Mutex lock(access_mutex());
+  is_referenced_ = true;
+  ref();
+}
+
+void NotificationChannel::Unref() {
+  ScopedLock::Mutex lock(access_mutex());
+  is_referenced_ = false;
+  if (!callback_pending_)
+    unref();
+}
+
+void NotificationChannel::Notify() {
+  {
+    ScopedLock::Mutex lock(access_mutex());
+    if (callback_pending_)
+      return;
+    callback_pending_ = true;
+    ref();
```

Comment:
It's not safe to call uv_ref() or uv_unref() from anywhere but the "owning" thread.